backend-tests/tests/test_reporting.py

```diff
@@ -0,0 +1,1225 @@
+
+    useradmm = ApiClient(useradm.URL_MGMT)
+    devauthd = ApiClient(deviceauth.URL_DEVICES)
+    devauthm = ApiClient(deviceauth.URL_MGMT)
+    invd = ApiClient(inventory.URL_DEV)
+
+    for inv in dev_inventories:
+        user = tenant.users[0]
+        utoken = useradmm.call(
+            "POST", useradm.URL_LOGIN, auth=(user.name, user.pwd)
+        ).text
+        assert utoken != ""
+
+        tenant.api_token = utoken
+        device = make_accepted_device(
+            devauthd, devauthm, utoken, tenant_token=tenant.tenant_token
+        )
+        tenant.devices.append(device)
+
+        attrs = dict_to_inventoryattrs(inv)
+        rsp = invd.with_auth(device.token).call(
+            "PATCH", inventory.URL_DEVICE_ATTRIBUTES, body=attrs
+        )
+        assert rsp.status_code == 200
+        device.inventory = inv
+
+    return tenant
+
+
+def add_devices_to_user(user, dev_inventories):
+    try:
+        user.devices
+    except AttributeError:
+        user.devices = []
+
+    useradmm = ApiClient(useradm.URL_MGMT)
+    devauthd = ApiClient(deviceauth.URL_DEVICES)
+    devauthm = ApiClient(deviceauth.URL_MGMT)
+    invd = ApiClient(inventory.URL_DEV)
+
+    for inv in dev_inventories:
+        utoken = useradmm.call(
+            "POST", useradm.URL_LOGIN, auth=(user.name, user.pwd)
+        ).text
+        assert utoken != ""
+
+        user.api_token = utoken
+        device = make_accepted_device(devauthd, devauthm, utoken)
+        user.devices.append(device)
+
+        attrs = dict_to_inventoryattrs(inv)
+        rsp = invd.with_auth(device.token).call(
+            "PATCH", inventory.URL_DEVICE_ATTRIBUTES, body=attrs
+        )
+        assert rsp.status_code == 200
+        device.inventory = inv
+
+    return user
+
+
+def maybe_list(v):
+    return v if type(v) == list else [v]
+
+
+@pytest.fixture(scope="function")
+def user(clean_mongo):
+    yield create_user("user-foo@acme.com", "correcthorse")
+
+
+@pytest.fixture(scope="function")
+def user_reporting(user):
+    add_devices_to_user(
+        user,
+        [
+            {"artifact": "v1", "py3": "3.3", "py2": "2.7", "idx": 0},
+            {"artifact": ["v1", "v2"], "py3": "3.5", "py2": "2.7", "idx": 1},
+            {"artifact": ["v1", "v2"], "py2": "2.6", "idx": 2},
+            {"artifact": ["v1", "v2"], "py2": "2.7", "idx": 3},
+            {"artifact": "v2", "py3": "3.5", "py2": "2.7", "idx": 4},
+            {"artifact": "v2", "py3": "3.6", "py2": "2.7", "idx": 5},
+            {"artifact": ["v2", "v3"], "py3": "3.6", "idx": 6},
+            {"artifact": ["v2", "v3"], "py3": "3.7", "idx": 7},
+            {"artifact": ["v2", "v3"], "py3": "3.7", "idx": 8},
+            {"artifact": ["v1", "v2", "v3"], "py3": "3.8", "idx": 9},
+        ],
+    )
+    # sleep a few seconds waiting for the data propagation to the reporting service
+    # and the Elasticsearch indexing to complete
+    time.sleep(reporting.REPORTING_DATA_PROPAGATION_SLEEP_TIME_SECS)
```

Comment:
Historically speaking, waiting for a fixed duration is a common source of spurious errors. Does it make sense to instead poll for the given number of devices to reach the ES index with a more relaxed time constraint (e.g. a minute or two)?

Reply:
I don't know, to be honest I liked the idea of waiting a fixed amount of time; if the reporting service is unable to catch up with the data within N seconds, then it is correct that the test fails, as this affects the UX. Does it then make sense to poll every 500ms up to 4 seconds?

Reply:
4 seconds might be loads of time on the local machine, but considering that this test is running on build hosts running multiple containerized jobs in parallel there might be some unexpected hiccups. If this is not the case, then ignore this comment.

Reply:
The theory says that we should not wait more than 2 seconds (the indexer runs every 1000ms, see [here](https://github.com/mendersoftware/reporting/blob/master/config.yaml#L65)); I doubled it to 4 seconds to be safe. Even under load, if it takes more we have an issue. I'm tempted to leave it as it is. We'll monitor the test results and see how it goes.